lib/ui/painting.dart

```diff
@@ -2855,11 +2858,45 @@ class ImageFilter {
       : assert(matrix4 != null), // ignore: unnecessary_null_comparison
         _data = Float64List.fromList(matrix4),
         _filterQuality = filterQuality,
+        _colorFilter = null,
+        _innerFilter = null,
+        _outerFilter = null,
         _type = _kTypeMatrix {
     if (matrix4.length != 16)
       throw ArgumentError('"matrix4" must have 16 entries.');
   }
 
+  /// Creates an image filter from a [ColorFilter].
+  ///
+  /// The resulting filter is functionally equivalent to the original
+  /// [ColorFilter], but can be used and composed like an [ImageFilter]. For
+  /// example, the [BackdropFilter] widget applies an [ImageFilter] to the
+  /// painted content. To use a [ColorFilter] in a [BackdropFilter], first
+  /// convert the [ColorFilter] to an [ImageFilter] using
+  /// [ImageFilter.fromColorFilter].
+  ///
+  /// See the [ColorFilter] class for available color filters.
+  ImageFilter.fromColorFilter(ColorFilter colorFilter)
```

Comment:
An thought on constructing these - rather than an ImageFilter constructor, what about `asImageFilter` on ColorFilter? Possibly a lazy getter?

`ImageFilter.fromColorFilter(myFilter)`
`ImageFilter.fromColorFilter(ColorFilter.xxx(...))`

vs.

`myFilter.asImageFilter;`
`ColorFilter.xxx(...).asImageFilter;`

Other `asFoo()` methods seem to be actual methods rather than getters, so the above should maybe have an `()` appended to them...

Reply:
Should I keep the ImageFilter.fromColorFilter constructor for discoverability?

Reply:
I believe toImageFilter would be more idiomatic

Reply:
I'm trying to wrap my head around those two idioms and I think we are in between them.

In one sense, we are talking about 2 separate objects so there is a "conversion" involved with respect to the type system.

But, under the covers, the conversion is mostly cosmetic. The same code is used, but we want to express that same operation as a different type so it can fit into more roles. It is very much as if we are putting "ImageFilter" clothing on a "ColorFilter" body.

Is this closer to ByteBuffer.as<Typed>List? Or Object.toString() or Iterable.toList()?'

In some sense, we could have ColorFilter subclass ImageFilter, but that might be conceptually unclear. Every color filter can be an image filter because it can be applied to every color in the image. A ColorFilter can always do the work expected of an ImageFilter and it can also filter a color without the color being associated with an Image as an additional capability.

A string doesn't really "do" the same things as the object it came from. A List can be manipulated unlike the Iterable it came from. A TypedList is really the same data as the ByteBuffer, but you interpret it differently.

I lean towards "asImageFilter".

Reply:
I don't have a strong opinion on whether to keep the ImageFilter constructor. I haven't looked around to see if there are other cases where we have Ainstance.toB() as well as B.from(Ainstance).

One glitch I see is that the ImageFilter that comes from a ColorFilter constructor prints out as "ImageFilter.fromColorFilter(ColorFilter.xxx(...))" in order to match its constructor's name, but it feels a little verbose for describing what it is. Is it really critical that objects with variants that come from various constructors mirror the names of those constructors? If we ditch the constructor, does that give us leverage for having the string form be something less crufty?

Could myColorFilter.asImageFilter().toString() yield "ColorFilter(...)" because of the as-ness of its origin? In other words, would it be jarring or desirable to have paint.imageFilter.toString() print out as a ColorFilter?

And if not, at least should the printout be more compact than "ImageFilter.fromColorFilter(ColorFilter(...))"? Perhaps "ImageFilter.color(...)" or maybe "ColorFilter.xxx(...).asImageFilter()"?

Reply:
Hmm.  Since `ImageFilter` has no public members, why not just make `ColorFilter` implement `ImageFilter`? We could just make a private member do the work we need right?

Reply:
I would be good with that.

Reply:
I should add, though, that I have reservations about the way that ColorFilter/ImageFilter are implemented as opaque objects. It would be nice to be able to say "what kind of ImageFilter is this?" at times, but there is no way to inspect any aspect of the object.

That makes it easy to do this "implements", but it is taking advantage of a lack of inspectability.

Until recently you couldn't even figure out what they did by printing them as they had no reasonable toString(), they were completely opaque.

Unfortunately, fixing this would involve turning CF and IF into abstract base classes and have all of their variants be concrete subclasses. I'm not sure we can go that route any more without a serious break in the API, though...

Reply:
(Though even in a world where IF was an abstract class, that would impose no real obstacles to having CF implement IF, so I guess that reservation doesn't really conflict with the suggestion here...)

Reply:
If we have a good use case for inspecting the type we could certainly expose something for it. But the user creating the type who needs it should be able to just know it right?

Reply:
I'm imagining something like a framework that a developer hands a filter to and the framework might want to make some optimizations depending on the filter. For example, a ColorFilter can't expand the bounds of the source, but a blur filter can. That's not a great example because in that case you could do an "is" test on CF, but potentially other optimizations might depend on information that isn't already exposed.

In any case, none of that would ever prevent CD from implementing the IF interface so I think we should take that suggestion...

Reply:
Tried this out today, making `ColorFilter` an `ImageFilter` would make it no longer const-constructable.

Reply:
I think that was me that cached the native image filter. ColorFilter just regenerates it on the fly every time. Does that end up being only once in practice if you use the const constructors?

I don't know what the value of making ColorFilter const is, but perhaps ImageFilter should mirror it in that sense?

@dnfield ?

Reply:
We don't want to have ColorFilter lose its const-ness. That'd be an unfortunate breaking change. 

Caching the native access offers some performance benefit too. It's not clear what would offer better advantages (being able to have a const ImageFilter with more native access, or caching the native field and losing const-ness) without a benchmark, but I honestly wouldn't spend a ton of time on that for this.

It's probably fine to either make ImageFilter const constructible and have it match the way ColorFilter behaves, or to just go with the to/from (or as) methods.

Reply:
If we ever find a difference in a benchmark we always have the option to non-lazily initialize the cached native filter even if the objects are const. I doubt apps create many filters that don't end up being passed to native anyway so I'm not sure why the creation of the native version is done lazily in the first place.

Reply:
Going with adding a `toImageFilter` method because `ImageFilter` can't be made const constructable (without introducing a breaking change). It stores its matrix in a `Float64List` (doesn't have a const constructor).

Reply:
We could store it in a double list, but that's ok.

I don't think we could non-lazily const construct the _ImageFilter class, because it involves calling some native methods.

Reply:
There's an `ImageFilter` constructor that takes a `Float64List` so changing the representation of the matrix and making the constructor const would be breaking I think

https://github.com/flutter/engine/blob/0cda2931e502809dcc8dcb335816b4259fbe04ba/lib/ui/painting.dart#L2853-L2861

Reply:
It could be turned into a factory that passes through to the const ctor that takes a List<double>

Reply:
Should this constructor that takes a `Float64List` be deprecated then?

Reply:
How efficient is `List<double>` for passing the matrix to a native method? I suppose we could have 16 arguments to the method and just unpack them in the native constructor call.

Is there a reason there is not const constructor for Float64List?

Reply:
I think we're going to convert the `List<double>` back to a `Float64List` when creating the native image filter (`_ImageFilter`), so there's going to be a round trip from `Float64List` to `List<double>` and then back to `Float64List`.

> Is there a reason there is not const constructor for Float64List?
I don't have an answer to that but it looks like `Float64List` is an abstract class that has different implementations for dartvm and js?
https://github.com/dart-lang/sdk/blob/87027c97a12a9926c2af8770a3d2a5cebc1767f3/sdk/lib/typed_data/typed_data.dart#L2127-L2255

Anyways even if it's possible to add a const constructor that would require an upstream change.

Reply:
I was able to make CF extends IF with the following change (that needs a bit of cleanup...)

It relaxes the non-null-ness of the internal _data field of ImageFilter, but since it overlaps in functionality with the _matrix field of ColorFilter it may be better to converge them instead. The point of the exercise, though, is that you don't need to make all of the constructors of ImageFilter be const just to have a single const constructor for ColorFilter to pass into.

```
diff --git a/lib/ui/painting.dart b/lib/ui/painting.dart
index c75585237..52f750ec2 100644
--- a/lib/ui/painting.dart
+++ b/lib/ui/painting.dart
@@ -2624,7 +2624,7 @@ class MaskFilter {
 ///
 /// Instances of this class are used with [Paint.colorFilter] on [Paint]
 /// objects.
-class ColorFilter {
+class ColorFilter extends ImageFilter {
   /// Creates a color filter that applies the blend mode given as the second
   /// argument. The source color is the one given as the first argument, and the
   /// destination color is the one from the layer being composited.
@@ -2636,7 +2636,7 @@ class ColorFilter {
       : _color = color,
         _blendMode = blendMode,
         _matrix = null,
-        _type = _TypeMode;
+        super._(_TypeMode);
 
   /// Construct a color filter that transforms a color by a 5x5 matrix, where
   /// the fifth row is implicitly added in an identity configuration.
@@ -2702,7 +2702,7 @@ class ColorFilter {
       : _color = null,
         _blendMode = null,
         _matrix = matrix,
-        _type = _TypeMatrix;
+        super._(_TypeMatrix);
 
   /// Construct a color filter that applies the sRGB gamma curve to the RGB
   /// channels.
@@ -2710,7 +2710,7 @@ class ColorFilter {
       : _color = null,
         _blendMode = null,
         _matrix = null,
-        _type = _TypeLinearToSrgbGamma;
+        super._(_TypeLinearToSrgbGamma);
 
   /// Creates a color filter that applies the inverse of the sRGB gamma curve
   /// to the RGB channels.
@@ -2718,12 +2718,11 @@ class ColorFilter {
       : _color = null,
         _blendMode = null,
         _matrix = null,
-        _type = _TypeSrgbToLinearGamma;
+        super._(_TypeSrgbToLinearGamma);
 
   final Color? _color;
   final BlendMode? _blendMode;
   final List<double>? _matrix;
-  final int _type;
 
   // The type of SkColorFilter class to create for Skia.
   static const int _TypeMode = 1; // MakeModeFilter
@@ -2838,6 +2837,8 @@ class _ColorFilter extends NativeFieldWrapperClass2 {
 ///  * [SceneBuilder.pushImageFilter], which is the low-level API for using
 ///    this class as a child layer filter.
 class ImageFilter {
+  const ImageFilter._(this._type) : _data = null, _filterQuality = null;
+
   /// Creates an image filter that applies a Gaussian blur.
   ImageFilter.blur({ double sigmaX = 0.0, double sigmaY = 0.0 })
       : assert(sigmaX != null), // ignore: unnecessary_null_comparison
@@ -2867,10 +2868,9 @@ class ImageFilter {
     return list;
   }
 
-  final Float64List _data;
+  final Float64List? _data;
   final FilterQuality? _filterQuality;
   final int _type;
-  _ImageFilter? _nativeFilter;
 
   // The type of SkImageFilter class to create for Skia.
   static const int _kTypeBlur = 0;   // MakeBlurFilter
@@ -2884,7 +2884,7 @@ class ImageFilter {
         && other._filterQuality == _filterQuality;
   }
 
-  _ImageFilter _toNativeImageFilter() => _nativeFilter ??= _makeNativeImageFilter();
+  _ImageFilter _toNativeImageFilter() => _makeNativeImageFilter();
 
   _ImageFilter _makeNativeImageFilter() {
     switch (_type) {
@@ -2904,7 +2904,7 @@ class ImageFilter {
   String toString() {
     switch (_type) {
       case _kTypeBlur:
-        return 'ImageFilter.blur(${_data[0]}, ${_data[1]})';
+        return 'ImageFilter.blur(${_data![0]}, ${_data![1]})';
       case _kTypeMatrix:
         return 'ImageFilter.matrix($_data, $_filterQuality)';
       default:
@@ -2926,7 +2926,7 @@ class _ImageFilter extends NativeFieldWrapperClass2 {
     : assert(creator != null), // ignore: unnecessary_null_comparison
       assert(creator._type == ImageFilter._kTypeBlur) {
     _constructor();
-    _initBlur(creator._data[0], creator._data[1]);
+    _initBlur(creator._data![0], creator._data![1]);
   }
   void _initBlur(double sigmaX, double sigmaY) native 'ImageFilter_initBlur';
 
@@ -2937,10 +2937,10 @@ class _ImageFilter extends NativeFieldWrapperClass2 {
   _ImageFilter.matrix(this.creator)
     : assert(creator != null), // ignore: unnecessary_null_comparison
       assert(creator._type == ImageFilter._kTypeMatrix) {
-    if (creator._data.length != 16)
+    if (creator._data!.length != 16)
       throw ArgumentError('"matrix4" must have 16 entries.');
     _constructor();
-    _initMatrix(creator._data, creator._filterQuality!.index);
+    _initMatrix(creator._data!, creator._filterQuality!.index);
   }
   void _initMatrix(Float64List matrix4, int filterQuality) native 'ImageFilter_initMatrix';
 
```

Reply:
Basically all of the implementation of both of these other than the constructors is private. We could turn ImageFilter itself into an empty class with 2 factories and have all of its implementation except for the single "_toNativeImageFilter()" accessor be private to an internal implementation class of ImageFilter...

```
diff --git a/lib/ui/painting.dart b/lib/ui/painting.dart
index c75585237..c9e4c467d 100644
--- a/lib/ui/painting.dart
+++ b/lib/ui/painting.dart
@@ -2624,7 +2624,7 @@ class MaskFilter {
 ///
 /// Instances of this class are used with [Paint.colorFilter] on [Paint]
 /// objects.
-class ColorFilter {
+class ColorFilter implements ImageFilter {
   /// Creates a color filter that applies the blend mode given as the second
   /// argument. The source color is the one given as the first argument, and the
   /// destination color is the one from the layer being composited.
@@ -2762,6 +2762,9 @@ class ColorFilter {
     }
   }
 
+  @override
+  _ImageFilter _toNativeImageFilter() => _ImageFilter.color(this);
+
   @override
   int get hashCode => hashValues(_color, _blendMode, hashList(_matrix), _type);
 
@@ -2837,27 +2840,21 @@ class _ColorFilter extends NativeFieldWrapperClass2 {
 ///    this class as a backdrop filter.
 ///  * [SceneBuilder.pushImageFilter], which is the low-level API for using
 ///    this class as a child layer filter.
-class ImageFilter {
+abstract class ImageFilter {
   /// Creates an image filter that applies a Gaussian blur.
-  ImageFilter.blur({ double sigmaX = 0.0, double sigmaY = 0.0 })
-      : assert(sigmaX != null), // ignore: unnecessary_null_comparison
-        assert(sigmaY != null), // ignore: unnecessary_null_comparison
-        _data = _makeList(sigmaX, sigmaY),
-        _filterQuality = null,
-        _type = _kTypeBlur;
+  factory ImageFilter.blur({ double sigmaX = 0.0, double sigmaY = 0.0 }) {
+    return _ImageFilterImpl._(_ImageFilterImpl._kTypeBlur, _makeList(sigmaX, sigmaY), null);
+  }
 
   /// Creates an image filter that applies a matrix transformation.
   ///
   /// For example, applying a positive scale matrix (see [Matrix4.diagonal3])
   /// when used with [BackdropFilter] would magnify the background image.
-  ImageFilter.matrix(Float64List matrix4,
-                     { FilterQuality filterQuality = FilterQuality.low })
-      : assert(matrix4 != null), // ignore: unnecessary_null_comparison
-        _data = Float64List.fromList(matrix4),
-        _filterQuality = filterQuality,
-        _type = _kTypeMatrix {
+  factory ImageFilter.matrix(Float64List matrix4,
+      { FilterQuality filterQuality = FilterQuality.low }) {
     if (matrix4.length != 16)
       throw ArgumentError('"matrix4" must have 16 entries.');
+    return _ImageFilterImpl._(_ImageFilterImpl._kTypeMatrix, matrix4, filterQuality);
   }
 
   static Float64List _makeList(double a, double b) {
@@ -2867,6 +2864,12 @@ class ImageFilter {
     return list;
   }
 
+  _ImageFilter _toNativeImageFilter();
+}
+
+class _ImageFilterImpl implements ImageFilter {
+  _ImageFilterImpl._(this._type, this._data, this._filterQuality);
+
   final Float64List _data;
   final FilterQuality? _filterQuality;
   final int _type;
@@ -2878,12 +2881,13 @@ class ImageFilter {
 
   @override
   bool operator ==(Object other) {
-    return other is ImageFilter
+    return other is _ImageFilterImpl
         && other._type == _type
         && _listEquals<double>(other._data, _data)
         && other._filterQuality == _filterQuality;
   }
 
+  @override
   _ImageFilter _toNativeImageFilter() => _nativeFilter ??= _makeNativeImageFilter();
 
   _ImageFilter _makeNativeImageFilter() {
@@ -2921,10 +2925,13 @@ class ImageFilter {
 class _ImageFilter extends NativeFieldWrapperClass2 {
   void _constructor() native 'ImageFilter_constructor';
 
+  _ImageFilter.color(this.creator);
+
   /// Creates an image filter that applies a Gaussian blur.
-  _ImageFilter.blur(this.creator)
-    : assert(creator != null), // ignore: unnecessary_null_comparison
-      assert(creator._type == ImageFilter._kTypeBlur) {
+  _ImageFilter.blur(_ImageFilterImpl creator)
+      : assert(creator != null), // ignore: unnecessary_null_comparison
+        assert(creator._type == _ImageFilterImpl._kTypeBlur),
+        creator = creator {
     _constructor();
     _initBlur(creator._data[0], creator._data[1]);
   }
@@ -2934,9 +2941,10 @@ class _ImageFilter extends NativeFieldWrapperClass2 {
   ///
   /// For example, applying a positive scale matrix (see [Matrix4.diagonal3])
   /// when used with [BackdropFilter] would magnify the background image.
-  _ImageFilter.matrix(this.creator)
-    : assert(creator != null), // ignore: unnecessary_null_comparison
-      assert(creator._type == ImageFilter._kTypeMatrix) {
+  _ImageFilter.matrix(_ImageFilterImpl creator)
+      : assert(creator != null), // ignore: unnecessary_null_comparison
+        assert(creator._type == _ImageFilterImpl._kTypeMatrix),
+        creator = creator {
     if (creator._data.length != 16)
       throw ArgumentError('"matrix4" must have 16 entries.');
     _constructor();
```

Reply:
If we're making `ImageFilter` abstract then I guess we might as well turn blur/matrix filters into separate classes? I remember Ian mentioned there's an overhead when you override a method in different classes. I'll try running our benchmarks locally and see if there's a performance impact.

Reply:
That would also clear up a bunch of the _type silliness in toString, ==, etc.

Reply:
I'm liking the latest push. I'll look into it in detail tomorrow. On a quick glance, I noticed that you are creating non-private nativeFilter fields in all the subclasses which should probably have an underscore for privacy.

Reply:
I ran the "backdrop_filter_perf" benchmark on my pixel 3 with and without that commit 10 times each in a row. It seems the medians of `average_frame_build_time_millis` are pretty close:

Before: 0.5140952380952382
After: 0.5140869565217391

Reply:
Looks great!